### PR TITLE
Migrate to futures 0.3 from 0.1

### PR DESCRIPTION
**Pull request by nrc**
~~And some minor refactoring. Based on #40, it is worth us reviewing and landing that first.~~

PTAL @sunxiaoguang

### CODE REVIEW

**Comment by nrc** [comment]
Rebased to remove conflict, merged the last three commits to make it easier to review

**Comment by nrc** [comment]
Rebased on to #42 (since I expect that to land first). PTAL at the second and third commits @Hoverbear and @sunxiaoguang

**Comment by Hoverbear** [comment]
This does cause us not to follow https://github.com/tikv/rfcs/pull/7/files#diff-3333e76d7edf3ed642c27ac78f1795a6R25 until these features are stabilized.

That said: The RFC isn't accepted.

I think it's fine. What do you think @sunxiaoguang?

**Comment by ice1000** [comment]
async/await! Cool!

File: README.md [hunk 1]
```diff
@@ -37,3 +43,31 @@ You can access the documentation on your machine by running the following in any
...
+
+## Toolchain versions
+
+To check what version of Rust you are using, run
+
+```bash
+rustc --version
+```
+
+you'll see something like `rustc 1.36.0-nightly (a784a8022 2019-05-09)` where the `1.36.0` is the toolchain version, and `nightly` is the channel (stable/beta/nightly). To install another toolchain use
```
Comment (Hoverbear) [inline README.md]: These sentences have inconsistent capitalization. ('To build', '(where `nightly`', 'you'll see')

File: README.md [hunk 2]
```diff
@@ -12,18 +12,24 @@ With this crate you can easily connect to any TiKV deployment, interact with it,
...
 # ...Your other dependencies...
-tikv-client = "~0.1"
+tikv-client = { git = "https://github.com/tikv/client-rust.git" }
 ```
 
+The client requires a Git dependency until we can [publish it](https://github.com/tikv/client-rust/issues/32).
+
+There are [examples](examples) which show how to use the client in a Rust program.
+
+The examples and documentation use async/await syntax. This is a new feature in Rust and is currently unstable. To use async/await you'll need to add the feature flag `#![]` to your crate and use a nightly compiler (see below).
```
Comment (Hoverbear) [inline README.md]:
```suggestion
The examples and documentation use async/await syntax. This is a new feature in Rust and is currently unstable. To use async/await you'll need to add the feature flag `#![feature(async_await, await_macro)]` to your crate and use a nightly compiler (see below).
```

File: src/lib.rs [hunk 3]
```diff
@@ -31,7 +31,7 @@
 //! [dependencies]
```
Comment (Hoverbear) [inline src/lib.rs]: A couple lines above this we say we support stable but we don't now!

Comment (nrc) [reply src/lib.rs]: Hmm, looks like we duplicate the README in the crate docs. IMO, it is better not to include installation instructions in the crates docs, just how to use the API. I've updated the PR with that change, what do you think?

**Pull request merged**